core/src/com/projectkorra/projectkorra/ability/CoreAbility.java

```diff
@@ -928,43 +937,92 @@ public List<Location> getLocations() {
-		}
-		this.attributeModifiers.get(attribute).get(priority).add(new StoredModifier(uuid, modificationType, value));
-		this.attributesModified = false;
 		return this;
 	}
 
+	@Deprecated
 	public CoreAbility setAttribute(final String attribute, final Object value) {
-		Validate.notNull(attribute, "attribute cannot be null");
-		Validate.notNull(value, "value cannot be null");
-		Validate.isTrue(ATTRIBUTE_FIELDS.containsKey(this.getClass()) && ATTRIBUTE_FIELDS.get(this.getClass()).containsKey(attribute), "Attribute " + attribute + " is not a defined Attribute for " + this.getName());
-		this.attributeValues.put(attribute, value);
-		this.attributesModified = false;
 		return this;
 	}
 
+	/**
+	 * Recalculate what the ability's attributes should be. This is called
+	 * whenever an ability is created, but should be called whenever you want an
+	 * ability to recalculate some of it's values. E.g. day turns to night, AvatarState
+	 * gets toggled, etc.
+	 */
+	public void recalculateAttributes() {
+		if (recalculatingAttributes) return; //Stop recursion if an addon does something wrong, e.g. calls recalculateAttributes inside the event
+
+		recalculatingAttributes = true;
+
+		if (!attributeValuesCached) { //Cache initial values
+			try {
+				for (AttributeCache cache : ATTRIBUTE_FIELDS.get(this.getClass()).values()) { //Get all attributes for this ability and cache initial values
+					cache.getInitialValues().put(this, cache.getField().get(this));
+				}
+			} catch (IllegalAccessException e) {
+				e.printStackTrace();
+			}
+			attributeValuesCached = true;
+		}
+
+		attribute_loop:
+		for (AttributeCache cache : ATTRIBUTE_FIELDS.get(this.getClass()).values()) {
+			String attribute = cache.getAttribute();
+			Object initialValue = cache.getInitialValues().get(this);
+
+			if (initialValue == null) {
+				ProjectKorra.log.severe("Initial value for " + attribute + " is null for " + this.getName() + "! Please report this to PK!");
+				continue;
+			}
+
+			AbilityRecalculateAttributeEvent event = new AbilityRecalculateAttributeEvent(this, attribute, initialValue);
+			Bukkit.getServer().getPluginManager().callEvent(event);
+
+			try {
+				for (AttributeModification mod : event.getModifications()) {
+					if (mod.getModifier() == AttributeModifier.SET) {
+						cache.getField().set(this, mod.getModification());
+						continue attribute_loop;
+					} else {
+						Number number = (Number) initialValue;
+						Number newValue = mod.getModifier().performModification(number, (Number) mod.getModification());
+						initialValue = newValue;
+					}
+				}
+
+				cache.getField().set(this, initialValue);
+			} catch (IllegalArgumentException | IllegalAccessException e) {
+				ProjectKorra.log.severe("Failed to recalculate attribute " + attribute + " for " + this.getName() + "!");
+				e.printStackTrace();
+			}
+		}
+
+		recalculatingAttributes = false;
+	}
+
+	@Deprecated
 	private void modifyAttributes() {
```

Comment:
Is it okay that this now does nothing? Do you think a warning or something might be beneficial?

Reply:
Hmm. Never thought of that. While it isn't a bad idea, it could get spammy in the console. So maybe it's best as is. A silent "no longer works" vs "loud spammy warnings"